applications/serial_lte_modem/src/slm_at_fota.c

```diff
@@ -0,0 +1,299 @@
+			NULL, 0);
+	if (err) {
+		LOG_ERR("nrf_setsockopt error: %d", errno);
+		nrf_close(fd);
+		return -errno;
+	}
+	while (true) {
+		err = nrf_getsockopt(fd, NRF_SOL_DFU, NRF_SO_DFU_OFFSET,
+			&offset, &len);
+		if (err < 0) {
+			k_sleep(K_SECONDS(1));
+		} else {
+			LOG_INF("Erase completed");
+			break;
+		}
+	}
+
+	nrf_close(fd);
+	return 0;
+}
+
+static int do_fota_start(int op, const char *file_uri, int sec_tag,
+			const char *apn)
+{
+	int ret = -EINVAL;
+	struct http_parser_url parser;
+	char schema[8];
+	char path[URI_PATH_MAX];
+	char hostname[URI_HOST_MAX];
+
+	http_parser_url_init(&parser);
+	ret = http_parser_parse_url(file_uri, strlen(file_uri), 0, &parser);
+	if (ret) {
+		LOG_ERR("Parse URL error");
+		return -EINVAL;
+	}
+	memset(schema, 0x00, 8);
+	if (parser.field_set & (1 << UF_SCHEMA)) {
+		strncpy(schema,
+			file_uri + parser.field_data[UF_SCHEMA].off,
+			parser.field_data[UF_SCHEMA].len);
+	} else {
+		LOG_ERR("Parse schema error");
+		return -EINVAL;
+	}
+	memset(path, 0x00, URI_PATH_MAX);
+	if (parser.field_set & (1 << UF_PATH)) {
+		strncpy(path,
+			file_uri + parser.field_data[UF_PATH].off,
+			parser.field_data[UF_PATH].len);
+	} else {
+		LOG_ERR("Parse path error");
+		return -EINVAL;
+	}
+
+	memset(hostname, 0x00, URI_HOST_MAX);
+	strncpy(hostname, file_uri,
+		strlen(file_uri) - parser.field_data[UF_PATH].len);
+
+	/* start HTTP(S) FOTA */
+	if (slm_util_cmd_casecmp(schema, SCHEMA_HTTPS)) {
+		if (sec_tag == INVALID_SEC_TAG) {
+			LOG_ERR("Missing sec_tag");
+			return -EINVAL;
+		}
+		ret = fota_download_start(hostname, path, sec_tag, apn, 0);
+	} else if (slm_util_cmd_casecmp(schema, SCHEMA_HTTP)) {
+		ret = fota_download_start(hostname, path, -1, apn, 0);
+	}
+
+	return ret;
+}
+
+static int do_fota_stop(void)
+{
+	int ret = 0;
+	return ret;
+}
+
+static void fota_dl_handler(const struct fota_download_evt *evt)
+{
+	switch (evt->id) {
+	case FOTA_DOWNLOAD_EVT_PROGRESS:
+		sprintf(rsp_buf, "#XFOTA: %d%% downloaded\r\n",
+				evt->progress);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		break;
+	case FOTA_DOWNLOAD_EVT_FINISHED:
+		sprintf(rsp_buf, "#XFOTA: downloaded, reset to apply.\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		break;
+	case FOTA_DOWNLOAD_EVT_ERASE_PENDING:
+		LOG_INF("FOTA_DOWNLOAD_EVT_ERASE_PENDING");
+		break;
+	case FOTA_DOWNLOAD_EVT_ERASE_DONE:
+		LOG_INF("FOTA_DOWNLOAD_EVT_ERASE_DONE");
+		break;
+	case FOTA_DOWNLOAD_EVT_ERROR:
+		sprintf(rsp_buf, "#XFOTA: download error.\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		break;
+
+	default:
+		break;
+	}
+}
+
+/**@brief handle AT#XFOTA commands
+ *  AT#XFOTA=<op>,<file_uri>[,<sec_tag>[,<apn>]]
+ *  AT#XFOTA? TEST command not supported
+ *  AT#XFOTA=?
+ */
+static int handle_at_fota(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+	uint16_t op;
+#if FOTA_FUTURE_FEATURE
+	static bool paused;
+#endif
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		if (at_params_valid_count_get(&at_param_list) < 1) {
```

Comment:
Shouldn't this be `<= 1`? AFAIR the command itself is also counted as AT param, and we need command and operation.

Reply:
Fixed